src/libraries/Common/src/System/IO/MemoryMappedFileInternal.cs

```diff
@@ -0,0 +1,167 @@
+        {
+            if (fileStream == null)
+            {
+                throw new ArgumentNullException(nameof(fileStream), SR.ArgumentNull_FileStream);
+            }
+
+            if (mapName != null && mapName.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_MapNameEmptyString);
+            }
+
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_PositiveOrDefaultCapacityRequired);
+            }
+
+            if (capacity == 0 && fileStream.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_EmptyFile);
+            }
+
+            if (access < MemoryMappedFileAccess.ReadWrite ||
+                access > MemoryMappedFileAccess.ReadWriteExecute)
+            {
+                throw new ArgumentOutOfRangeException(nameof(access));
+            }
+
+            if (access == MemoryMappedFileAccess.Write)
+            {
+                throw new ArgumentException(SR.Argument_NewMMFWriteAccessNotAllowed, nameof(access));
+            }
+
+            if (access == MemoryMappedFileAccess.Read && capacity > fileStream.Length)
+            {
+                throw new ArgumentException(SR.Argument_ReadAccessWithLargeCapacity);
+            }
+
+            if (inheritability < HandleInheritability.None || inheritability > HandleInheritability.Inheritable)
+            {
+                throw new ArgumentOutOfRangeException(nameof(inheritability));
+            }
+
+            // flush any bytes written to the FileStream buffer so that we can see them in our MemoryMappedFile
+            fileStream.Flush();
+
+            long updatedCapacity = (capacity == DefaultSize) ? fileStream.Length : capacity;
+
+            // one can always create a small view if they do not want to map an entire file
+            if (fileStream.Length > updatedCapacity)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_CapacityGEFileSizeRequired);
+            }
+
+            return updatedCapacity;
+        }
+
+        internal static void VerifyParametersCreateNew(
+            string? mapName,
+            long capacity,
+            MemoryMappedFileAccess access,
+            MemoryMappedFileOptions options,
+            HandleInheritability inheritability)
+        {
+            if (mapName != null && mapName.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_MapNameEmptyString);
+            }
+
+            if (capacity <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_NeedPositiveNumber);
+            }
+
+            if (IntPtr.Size == 4 && capacity > uint.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_CapacityLargerThanLogicalAddressSpaceNotAllowed);
```

Comment:
Aside - feel free to ignore here. In general, it is nice for strings describing invalid arguments to include the value itself. I have found this can help to diagnose issues when you are presented with only the exception message (such as from a server log).

Among the strings like this are:
```
  <data name="ArgumentOutOfRange_CapacityLargerThanLogicalAddressSpaceNotAllowed" xml:space="preserve">
    <value>The capacity cannot be greater than the size of the system's logical address space.</value>
  </data>
  <data name="ArgumentOutOfRange_FileLengthTooBig" xml:space="preserve">
    <value>Specified file length was too large for the file system.</value>
  </data>
  <data name="ArgumentOutOfRange_NeedPositiveNumber" xml:space="preserve">
    <value>A positive number is required.</value>
  </data>
  <data name="ArgumentOutOfRange_PositiveOrDefaultCapacityRequired" xml:space="preserve">
    <value>The capacity must be greater than or equal to 0. 0 represents the size of the file being mapped.</value>
  </data>
  <data name="ArgumentOutOfRange_PositiveOrDefaultSizeRequired" xml:space="preserve">
    <value>The size must be greater than or equal to 0. If 0 is specified, the view extends from the specified offset to the end of the file mapping.</value>
  </data>
  <data name="ArgumentOutOfRange_CapacityGEFileSizeRequired" xml:space="preserve">
    <value>The capacity may not be smaller than the file size.</value>
  </data>
```

This wasn't generally historical practice in the BCL so there are many of these. But going forward (again not necessarily in this PR) it would be nice to improve as we go along.

Reply:
Note that one way of doing this is to use the overloads that take a value - they append the bad value to the message you supply. I personally prefer the value in the string as sometimes the two lines of the message get separated in the log, but that's not a big deal.

Reply:
Just noticed this comment. Thanks for your suggestion. I will address it on the next iteration.